File: docs/experiments/02-compile-array-fusion/compile_array_fusion.m
(* ==========================================================================
   Experiment 2 -- Compiled machine arrays and fused elementwise loops
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file compile_array_fusion.m
       Mathematica   wolframscript -file compile_array_fusion.m
       Python        python3 compile_array_fusion.py

   WHAT IT MEASURES.  An elementwise expression over a 10^6-element vector,
   written three ways: interpreted, compiled, and compiled with the loop
   FUSED.  Fusion is the whole subject -- `v^2 + 2 v + 1` unfused is four
   passes over 8 MB with three temporaries; fused it is one pass with none.

   The Sin/Exp/Sqrt row is deliberately included because it is libm-bound
   rather than memory-bound, so it shows what fusion is worth when the
   arithmetic, not the traffic, is the cost -- less, and the file says so.
   ========================================================================== *)

SetAttributes[bench, HoldRest];

(* bench[label, expr] -- one untimed warm-up, then the MINIMUM of three timed
   runs.  The minimum, not the mean: we are measuring the cost of the work, and
   every source of noise on a loaded machine can only add. *)
bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];


n = 10^6;
v = RandomReal[{0, 1}, n];

(* Compiled, fused: one pass, no temporaries. *)
f1 = Compile[{{a, _Real, 1}}, a^2 + 2 a + 1];
f2 = Compile[{{a, _Real, 1}}, Total[a^2 + 2 a + 1]];
f3 = Compile[{{a, _Real, 1}}, Total[a + a a]];
f4 = Compile[{{a, _Real, 1}}, Total[Sin[a] Exp[-a] + Sqrt[a]]];
f5 = Compile[{{a, _Real, 1}}, Sqrt[a] + a^2];
f6 = Compile[{{a, _Real, 1}}, Sin[a] Exp[-a] + Sqrt[a]];

Print["Experiment 2 -- compiled machine arrays and fused elementwise loops"];
Print[""];
Print["-- the same expression, interpreted then compiled --"];

bench["v^2 + 2 v + 1            (interpreted)", v^2 + 2 v + 1];
bench["v^2 + 2 v + 1            (compiled, fused)", f1[v]];
bench["Total[v^2 + 2 v + 1]     (interpreted)", Total[v^2 + 2 v + 1]];
bench["Total[v^2 + 2 v + 1]     (compiled, fused)", f2[v]];
bench["Total[v + v v]           (compiled, fused)", f3[v]];
bench["Sqrt[v] + v^2            (compiled, fused)", f5[v]];

Print[""];
Print["-- libm-bound bodies: fusion is worth less, because the cost is not traffic --"];
bench["Sin[v] Exp[-v] + Sqrt[v] (interpreted)", Sin[v] Exp[-v] + Sqrt[v]];
bench["Sin[v] Exp[-v] + Sqrt[v] (compiled, fused)", f6[v]];
bench["Total[Sin[v] Exp[-v] + Sqrt[v]] (compiled)", f4[v]];

Print[""];
Print["-- the elementwise primitives, for scale --"];
bench["Sin[v]",  Sin[v]];
bench["Exp[v]",  Exp[v]];
bench["v + 3. v", v + 3. v];

check["Total[v^2 + 2 v + 1] agrees", Round[f2[v] - Total[v^2 + 2 v + 1], 1.*^-6]];

File: docs/experiments/02-compile-array-fusion/compile_array_fusion.py
#!/usr/bin/env python3
"""Experiment 2 -- Compiled machine arrays and fused elementwise loops.

Runs the same expressions as ``compile_array_fusion.m``.  See ``README.md``.

    python3 compile_array_fusion.py

WHAT THE NUMPY COLUMN IS HERE.  NumPy is the UNFUSED reference: ``a**2 + 2*a +
1`` in NumPy is four passes over 8 MB with three temporaries, exactly like the
interpreted CAS form, because NumPy has no fusion either.  That is what makes
it the right control for this experiment -- the compiled-and-fused CAS numbers
should beat it, and the point of the row is by how much.

(``numexpr`` and ``numba`` do fuse; neither is installed on this host, and
substituting one would be comparing a different library.)
"""

import time

import numpy as np


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))

N = 10 ** 6
v = np.random.rand(N)


def main():
    print("Experiment 2 -- compiled machine arrays and fused elementwise loops")
    print("")
    print("-- NumPy is UNFUSED: each of these is several passes --")
    bench("v**2 + 2*v + 1", lambda: v ** 2 + 2 * v + 1)
    bench("(v**2 + 2*v + 1).sum()", lambda: (v ** 2 + 2 * v + 1).sum())
    bench("(v + v*v).sum()", lambda: (v + v * v).sum())
    bench("sqrt(v) + v**2", lambda: np.sqrt(v) + v ** 2)

    print("")
    print("-- libm-bound bodies --")
    bench("sin(v)*exp(-v) + sqrt(v)",
          lambda: np.sin(v) * np.exp(-v) + np.sqrt(v))
    bench("(sin(v)*exp(-v) + sqrt(v)).sum()",
          lambda: (np.sin(v) * np.exp(-v) + np.sqrt(v)).sum())

    print("")
    print("-- the elementwise primitives, for scale --")
    bench("sin(v)", lambda: np.sin(v))
    bench("exp(v)", lambda: np.exp(v))
    bench("v + 3.0*v", lambda: v + 3.0 * v)

    check("(v**2 + 2*v + 1).sum()", round(float((v ** 2 + 2 * v + 1).sum()), 6))


if __name__ == "__main__":
    main()
